Add support for custom chunk render types

FiniteReality opened the pull request:
This is the minimal set of changes needed to allow registration of custom render types to be used when rendering static level geometry. I've loosely based these changes on #8331 (such as including its mipmap fixes) but does not include many of the other changes made in that PR, such as:
- Using the TierSortingRegistry to sort and manage RenderTypes
- Supporting transparent/tripwire geometry - this is because I expect that most developers would want the translucent sort to be applied for geometry that is rendered around the same time as translucent geometry and that would require a larger set of changes to decide whether to enable the translucent sort or not without harming performance too much
- A configuration flag to use the vanilla codepath: the Forge codepath is always enabled as disabling it will cause holes in the world
- The additional types for registering a callback before rendering chunk geometry used for uploading uniforms and the like
- Updating MultiLayerModel as it has been replaced by CompositeModel, which I believe should already handle these changes if modders are registering their RenderType-s properly.

As with the other PR, the additional runtime cost is roughly the same, scaling with the same parameters.

Included is a test mod which registers a block which uses a shader to render each face as its normals. It can be accessed using the setblock command.

FiniteReality committed:
Add support for custom chunk render types

This is the minimal set of changes needed to allow registration of custom
render types to be used when rendering static level geometry. I've
loosely based these changes on #8331 (such as including its mipmap fixes)
but does not include many of the other changes made in that PR, such as:
- Using the TierSortingRegistry to sort and manage RenderTypes
- Supporting transparent/tripwire geometry - this is because I expect
  that most developers would want the translucent sort to be applied for
  geometry that is rendered around the same time as translucent geometry
  and that would require a larger set of changes to decide whether to
  enable the translucent sort or not without harming performance too
  much
- A configuration flag to use the vanilla codepath: the Forge codepath is
  always enabled as disabling it will cause holes in the world
- The additional types for registering a callback before rendering chunk
  geometry used for uploading uniforms and the like
- Updating MultiLayerModel as it has been replaced by CompositeModel,
  which I believe should already handle these changes if modders are
  registering their RenderType-s properly.

As with the other PR, the additional runtime cost is roughly the same,
scaling with the same parameters.

Included is a test mod which registers a block which uses a shader to
render each face as its normals. It can be accessed using the setblock
command.

marchermans commented:
I am going to be honest.

The state of the added features is not sufficient to be able to use this properly.
From what I know discussing that pull request back then, it is that for the functionality to be anywhere close to complete, transparency needs to be supported, which brings the removed points 1 and 2 with them.

Then to be able to do custom geometry uniforms properly are needed for shader expansion, which means point 3 is needed (although that could be spun off into a second PR)

What was the justification for not supporting transparency sorting?

FiniteReality commented:
> What was the justification for not supporting transparency sorting?

Mainly performance - doing the transparency sort currently is hardcoded to `RenderType.translucent()` and changing it currently means doing a linear lookup every time `renderChunkLayer` is called. Ideally, whether the translucent sort is needed or not would be defined by RenderType itself to avoid a linear lookup or a hardcoded check, but I felt it was best to start with the minimal changes necessary and elicit feedback over whether it was necessary at all.

marchermans commented:
> > What was the justification for not supporting transparency sorting?
> 
> Mainly performance - doing the transparency sort currently is hardcoded to `RenderType.translucent()` and changing it currently means doing a linear lookup every time `renderChunkLayer` is called. Ideally, whether the translucent sort is needed or not would be defined by RenderType itself to avoid a linear lookup or a hardcoded check, but I felt it was best to start with the minimal changes necessary and elicit feedback over whether it was necessary at all.

To be able to merge this, I feel like we should convert it into a draft, and I would urge you to research ways of handling transparency sorting properly. In other words, find a nice API that circumvents the need for the linear look up. Something I was thinking about was the following: If we reintroduce the tier sorting registry, we kind of know which elements do need tier sorting or not. Alternatively you can for example expand the Composite State Builder to support it, and pass the information along that way into the rendertype?

FiniteReality commented:
> Alternatively you can for example expand the Composite State Builder to support it, and pass the information along that way into the rendertype?

This is the solution I'm currently working on now - I just need to make sure I have the EXC entries correct and then I'll push my changes.

FiniteReality committed:
Add support for transparent geometry sorting

FiniteReality committed:
Re-disable the test mod I left enabled by accident

XFactHD commented on patches/minecraft/net/minecraft/client/renderer/LevelRenderer.java.patch:
```diff
@@ -132,7 +136,22 @@
++         }
+          profilerfiller.m_6182_("string");
+-         this.m_172993_(RenderType.m_110503_(), p_109600_, d0, d1, d2, p_254120_);
++         // FORGE: render all translucent layers in registration order
```
These are tripwire layers, not translucent layers

XFactHD commented on patches/minecraft/net/minecraft/client/renderer/LevelRenderer.java.patch:
```diff
@@ -142,9 +161,23 @@
           profilerfiller.m_6182_("string");
-          this.m_172993_(RenderType.m_110503_(), p_109600_, d0, d1, d2, p_254120_);
+-         this.m_172993_(RenderType.m_110503_(), p_109600_, d0, d1, d2, p_254120_);
++         // FORGE: render all translucent layers in registration order
```
These are tripwire layers, not translucent layers

XFactHD commented on src/main/java/net/minecraftforge/client/ChunkBufferLayerManager.java:
```diff
@@ -0,0 +1,127 @@
+    /**
+     * Returns the list of all registered solid chunk buffer layers.
+     *
+     * @return the list of all registered solid chunk buffer layers.
```
This returns translucent layers, not solid layers

XFactHD commented on src/main/java/net/minecraftforge/client/ChunkBufferLayerManager.java:
```diff
@@ -0,0 +1,127 @@
+    /**
+     * Returns the list of all registered solid chunk buffer layers.
+     *
+     * @return the list of all registered solid chunk buffer layers.
```
This returns tripwire layers, not solid layers

XFactHD commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -5,6 +5,7 @@
 
 package net.minecraftforge.client;
 
+import com.google.common.collect.ImmutableList;
```
Unnecessary import changes

XFactHD commented on src/test/resources/assets/custom_chunk_layer_test/models/block/test.json:
```diff
@@ -0,0 +1,17 @@
+      "from": [ 0, 0, 0 ],
+      "to": [ 16, 16, 16 ],
+      "faces": {
+        "down":  { "texture": "#missing", "cullface": "down" },
```
This should ideally use a placeholder texture to avoid resource loading errors. The `forge:white` texture shipped in Forge should be fine for this

FiniteReality replied:
Surprisingly I didn't seem to get any while testing the model, so I left it - I have fixed it though.

XFactHD reviewed (commented):
Just a few minor nitpicks here and there. One point, which I also mentioned on ama's original PR, that needs discussion in my opinion is the map of so-called "fixed layers". Quoting my original comment from ama's PR (https://github.com/MinecraftForge/MinecraftForge/pull/8331/files#r872758451):

> In my opinion this [the patch of `RenderType.chunkBufferLayers()`] is missing an accompanying patch in `RenderBuffers`: The `RenderBuffers#fixedBuffers` map which is used to construct the `RenderBuffers#bufferSource` needs to contain all chunk layers because requesting a buffer from said buffer source for a certain `RenderType` after using a different `RenderType` that isn't in the map of fixed buffers will cause the current data in the buffer to be uploaded to the GPU because "non-fixed" `RenderType`s use a shared buffer.
>
> While this is not an issue for the static chunk rendering, which uses the `ChunkBufferBuilderPack` to get its buffers, I would still expect a `RenderType`, which has a pre-allocated buffer and can be used in static chunk rendering, to be useable in a dynamic context (i.e. `EntityRenderer` or `BlockEntityRenderer`) without causing an upload to the GPU every time I switch the `RenderType` in use to render the next thing in the dynamic renderer (which kills performance when a few instances of the related `Entity` or `BlockEntity` are present).

There needs to be some way to add custom chunk render types to the `RenderBuffers#fixedBuffers` map. The difficulty with this is that the use of this map in `MultiBufferSource.BufferSource#endBatch()` is sensitive to the order of the entries, so a way would need to be found to insert between the vanilla entries (which the `Object2ObjectLinkedOpenHashMap` used there doesn't seem to support.

A nice side-effect of an API for this would be the ability to make custom glint render types since these must be in that map as well due to glint rendering using two different buffers at the same time, which only works properly if the first render type (the glint type in case of the `ItemRenderer`) is a fixed type.

FiniteReality committed:
Fix typos in comments, use forge:white texture

FiniteReality commented:
> I would still expect a `RenderType`, which has a pre-allocated buffer and can be used in static chunk rendering, to be useable in a dynamic context (i.e. `EntityRenderer` or `BlockEntityRenderer`) without causing an upload to the GPU every time I switch the `RenderType` in use to render the next thing in the dynamic renderer (which kills performance when a few instances of the related `Entity` or `BlockEntity` are present).

The render types used in static chunk rendering are not usable in a dynamic context; they use a completely different vertex format and the parameters are different. I do agree that when registering a rendertype for static rendering that you would most likely want to register one for dynamic rendering too, however - this is where NamedRenderTypeManager comes in, as it is a mapping of static->dynamic render types. Are you suggesting that this event should append to the `fixedBuffers` map too, somewhere between `solidBlockSheet`, `cutoutBlockSheet` and `translucentCullBlockSheet`?

FiniteReality committed:
Fix item rendering issue

FiniteReality committed:
Fix exc errors

XFactHD commented:
> The render types used in static chunk rendering are not usable in a dynamic context; they use a completely different vertex format and the parameters are different.

That's a fair point, I worded that poorly.

> Are you suggesting that this event should append to the `fixedBuffers` map too, somewhere between `solidBlockSheet`, `cutoutBlockSheet` and `translucentCullBlockSheet`?

While that would probably work fine, I feel like it would be too limiting. In my opinion it should either be a separate API entirely or the event should call into a separate API to also allow modders to add other custom render types (like the glint stuff I mentioned) to the `fixedBuffers` map, which aren't chunk render types.

FiniteReality commented:
> While that would probably work fine, I feel like it would be too limiting. In my opinion it should either be a separate API entirely or the event should call into a separate API to also allow modders to add other custom render types (like the glint stuff I mentioned) to the `fixedBuffers` map, which aren't chunk render types.

Since managing buffers affects both static and dynamic render types, it may be worthwhile to merge this new event and `RegisterNamedRenderTypesEvent`, or otherwise supersede both of them with a unified event which handles registering and ordering render types. Alternatively, we can keep the two events and one can be for registering the desired order of both static and dynamic rendering, while the other can be for registering the static/dynamic render types by name.

Alternatively, a new buffer manager can be developed which manages both static and dynamic buffers, in the correct order. I think this might be the way to go if we want something maintainable long-term, even if it does mean more patches to Vanilla code.

FiniteReality committed:
Extract buffer management into a new manager

Breaking changes in this commit:
- RegisterNamedRenderTypesEvent is no longer fired
  - A new event with the same purpose has been added

Definitely not done here, but it runs stably and should allow initial
review. Whitespace changes to come, of course.

FiniteReality committed:
Actually use the new codepath for batched geometry

FiniteReality committed:
Fix license header violations

FiniteReality commented:
Okay, so far I've tested with vanilla code and everything seems to work. The compatibility check is failing because I removed NamedRenderTypeManager.init (which is marked ApiStatus.Internal) and changed ClientForgeMod.onRegisterNamedRenderTypes to use the new event. Should I keep the originals but mark them `@Deprecated`?

I'm going to move to testing with some mods that play around with rendering now, as it appears vanilla items and entities appear to be rendering correctly.

FiniteReality committed:
Merge 1.19.4 changes in

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=9366) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=9366) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/MinecraftForge/MinecraftForge?pullRequest=9366) it.</sub>
